### PR TITLE
feat: add sdxl sampler

### PR DESCRIPTION
**Pull request by Fzilan**
add HeunEDMSampler, DPMPP2MSampler

### CODE REVIEW

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py [hunk 1]
```diff
@@ -126,3 +126,89 @@ def __call__(self, model, x, cond, uc=None, num_steps=None):
 class EulerEDMSampler(EDMSampler):
     def possible_correction_step(self, euler_step, x, d, dt, next_sigma, model, cond, uc):
         return euler_step
+
+
+class HeunEDMSampler(EDMSampler):
+    def possible_correction_step(self, euler_step, x, d, dt, next_sigma, denoiser, cond, uc):
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py:132]: denoiser -> model

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py [hunk 2]
```diff
@@ -126,3 +126,89 @@ def __call__(self, model, x, cond, uc=None, num_steps=None):
...
+        else:
+            denoised_d = mult[2] * denoised - mult[3] * old_denoised
+            x_advanced = mult[0] * x - mult[1] * denoised_d
+
+            # apply correction if noise level is not 0 and not first step
+            x = ops.where(append_dims(next_sigma, x.ndim) > 0.0, x_advanced, x_standard)
+
+        return x, denoised
+
+    def __call__(self, denoiser, x, cond, uc=None, num_steps=None, **kwargs):
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py:198]: denoiser -> model

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py [hunk 3]
```diff
@@ -126,3 +126,89 @@ def __call__(self, model, x, cond, uc=None, num_steps=None):
...
+        x, s_in, sigmas, num_sigmas, cond, uc = self.prepare_sampling_loop(x, cond, uc, num_steps)
+
+        old_denoised = None
+        for i in self.get_sigma_gen(num_sigmas):
+            x, old_denoised = self.sampler_step(
+                old_denoised,
+                None if i == 0 else s_in * sigmas[i - 1],
+                s_in * sigmas[i],
+                s_in * sigmas[i + 1],
+                denoiser,
```
Comment (zhanghuiyao) [inline examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py:208]: denoiser -> model

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py [hunk 4]
```diff
@@ -126,3 +126,89 @@ def __call__(self, model, x, cond, uc=None, num_steps=None):
...
+        cond,
+        uc=None,
+    ):
+        denoised = self.denoise(x, denoiser, sigma, cond, uc)
+
+        h, r, t, t_next = self.get_variables(sigma, next_sigma, previous_sigma)
+        mult = [append_dims(mult, x.ndim) for mult in self.get_mult(h, r, t, t_next, previous_sigma)]
+
+        x_standard = mult[0] * x - mult[1] * denoised
+        if old_denoised is None or ops.sum(next_sigma) < 1e-14:
```
Comment (SamitHuang) [inline examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py:186]: Why do we need to judge `ops.sum(next_sigma) < 1e-14` ?

**Review by zhanghuiyao — APPROVED** [review]

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py (hunk 1 above)
Comment (Fzilan) [reply examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py]: fixed

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py (hunk 2 above)
Comment (Fzilan) [reply examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py]: fixed

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py (hunk 3 above)
Comment (Fzilan) [reply examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py]: fixed

File: examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py (hunk 4 above)
Comment (Fzilan) [reply examples/stable_diffusion_xl/gm/modules/diffusionmodules/sampler.py:186]: referrence to implement in:
https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py
https://github.com/Stability-AI/generative-models/blob/main/sgm/modules/diffusionmodules/sampling.py

**Review by SamitHuang — APPROVED** [review]

**Pull request merged**